builder_test.go

```diff
@@ -0,0 +1,242 @@
+)
+
+func ParseHschema(file string) (*hschema.HyperSchema, error) {
+	b, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, fmt.Errorf("fail to read the source JSON Schema file: %s", err)
+	}
+
+	var m map[string]interface{}
+	if err := yaml.Unmarshal(b, &m); err != nil {
+		return nil, fmt.Errorf("fail to unmarshal YAML: %s", err)
+	}
+
+	hs := hschema.New()
+	if err := hs.Extract(m); err != nil {
+		return nil, fmt.Errorf("fail to extract JSON Schema: %s", err)
+	}
+
+	return hs, nil
+}
+
+func TestBuilderLoopRef(t *testing.T) {
+	hs, err := ParseHschema("./test/ref_loop.yml")
+	if err != nil {
+		t.Fatal(err)
+	}
+	b := NewBuilder()
+	_, err = b.Build(hs)
+	if err == nil {
+		t.Fatalf("build should be failed: %+v", b)
+	}
+}
+
+func TestBuillderNotHaveHref(t *testing.T) {
+	hs, err := ParseHschema("./test/has_not_href.yml")
+	if err != nil {
+		t.Fatal(err)
+	}
+	b := NewBuilder()
+	ts, err := b.Build(hs)
+	if err != nil {
+		t.Fatalf("fail to build: %s", err)
+	}
+
+	if ts.HyperSchema.Title != "has not href" {
+		t.Error("fail to get title")
+	}
+}
+
+func TestBuilderPassBuild(t *testing.T) {
+	hs, err := ParseHschema("./test/pass.yml")
+	if err != nil {
+		t.Fatal(err)
+	}
+	b := NewBuilder()
+	ts, err := b.Build(hs)
+	if err != nil {
+		t.Fatalf("fail to build: %s", err)
+	}
+
+	if len(ts.Objects) != 1 {
+		t.Fatal("fail to parse Object type schema")
+	}
+
+	if ts.Objects[0] == nil || ts.Objects[0].Title() != "test object" {
+		t.Errorf("fail to get Object type schema: %+v", ts.Objects[0])
+	}
+
+	if len(ts.Arrays) != 1 {
+		t.Fatal("fail to parse Array type schema")
+	}
+
+	if ts.Arrays[0] == nil || ts.Arrays[0].Title() != "test array" {
+		t.Errorf("fail to get Arrays type schema: %+v", ts.Arrays[0])
+	}
+
+	if len(ts.Booleans) != 1 {
+		t.Fatal("fail to parse Boolean type schema")
+	}
+
+	if ts.Booleans[0] == nil || ts.Booleans[0].Title() != "test boolean" {
+		t.Errorf("fail to get Boolean type schema: %+v", ts.Booleans[0])
+	}
+
+	if len(ts.Numbers) != 1 {
+		t.Fatal("fail to parse Number type schema")
+	}
+
+	if ts.Numbers[0] == nil || ts.Numbers[0].Title() != "test number" {
+		t.Errorf("fail to get Numbers type schema: %+v", ts.Numbers[0])
+	}
+
+	if len(ts.Integers) != 2 {
+		t.Fatal("fail to parse Integer type schema")
+	}
+
+	if ts.Integers[0] == nil || ts.Integers[0].Title() != "test integer" {
+		t.Errorf("fail to get Integer type schema: %+v", ts.Integers[0])
+	}
+
+	if len(ts.Strings) != 1 {
+		t.Fatal("fail to parse String type schema")
+	}
+
+	if ts.Strings[0] == nil || ts.Strings[0].Title() != "test string" {
+		t.Errorf("fail to get Strings type schema: %+v", ts.Strings[0])
+	}
+
+	if len(ts.Properties) != 2 {
+		t.Fatal("fail to parse Properties type schema")
+	}
+
+	if len(ts.Definitions) != 7 {
+		t.Fatal("fail to parse Definitions")
+	}
+
+	for _, v := range ts.Properties {
+		switch v.Key() {
+		case "test_multitype":
+			i, ok := v.(*jstypes.Integer)
+			if ok != true {
```

Comment:
@meta-closure 

You should shorten `ok != true` to `!ok`.
I found 11 same cases in this file.